### PR TITLE
API New SSNotice JS tools

### PR DESCRIPTION
**Pull request by colymba**
Worked on replacing `jquery.notice.js` with a custom set of function and streamline the look of notice/messages in the CMS. The old one would usually get in the way for me, not be very useful (for errors) and look out of place. See screenshots below.

This basically create a library in the `ss` namespace `ss.notice` with a few function for displaying messages, notices, errors, warning. It would be great if someone could look at it and give some opinion/pointers.

* Goes in pair with CMS pull request (https://github.com/silverstripe/silverstripe-cms/pull/1024)
* Left the old `statusMessage` and `errorMessage` function with deprecation notices
* Should message + notice disappear automatically or stay by default like errors wand warnings?
* Multiple instance of empty `<p>` message tag in various templates which I could not find use for anywhere so they were deleted.
* Behat test updated, although I can't get Behat to run on my system so this might need some reviewing

Before:
image

image

After:
image

image

image

### CODE REVIEW

File: admin/javascript/SSNotice.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+(function($, root) {
+
+	// local working version of the library object
+	var lib = {};
+
+	/**
+	 * Library settings
+	 */
+	lib.settings = {
+		$parent:  $('.cms-content-fields'),
```
Comment (tractorcow) [inline admin/javascript/SSNotice.js]: Should the setting rather be a css selector, which is not evaluated until after lib.show is invoked? That might behave a bit more robustly.

**Comment by tractorcow** [comment]
Sorry for not looking after this PR; but if you want to rebase and squash again maybe we should get this into master. :) thanks @colymba

**Comment by colymba** [comment]
Change `$parent` to a CSS selector and squashed the PR... should be good to go...

**Comment by tractorcow** [comment]
Hm, it looks like there are some serious behat errors now?

File: admin/javascript/LeftAndMain.js [hunk 2]
```diff
@@ -139,8 +139,6 @@ jQuery.noConflict();
 
 			// Handle custom status message headers
 			var msg = (xhr.getResponseHeader('X-Status')) ? xhr.getResponseHeader('X-Status') : xhr.statusText,
-				reathenticate = xhr.getResponseHeader('X-Reauthenticate'),
```
Comment (tractorcow) [inline admin/javascript/LeftAndMain.js]: This is necessary for the cms re-authentication feature.

**Comment by chillu** [comment]
> Default behaviour is message and success notices disappear automatically after 10s

Sorry for the very delayed feedback on this! This will move shift fields upwards by ~30px, right? Not very user friendly, both if this coincides with the user trying to click small items like the "edit url" button, or if they're working in the rich text editor and trying to place cursor focus. Particularly since this new styling applies to notice behaviours after every save, I could see it becoming quite annoying to users. @clarkepaul @stevie-mayhew Thoughts?

File: admin/javascript/SSNotice.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+	// local working version of the library object
+	var lib = {};
+
+	/**
+	 * Library settings
+	 */
+	lib.settings = {
+		parentSelector: '.cms-content-fields',
+		$notice: $('<p class="message"/>'),
```
Comment (dhensby) [inline admin/javascript/SSNotice.js]: would it not be better to have this in the template already? we're mixing concerns by placing HTML in the JS file

**Comment by stevie-mayhew** [comment]
I think that an error or a notification should stay in place until the user dismisses it - automatically hiding something can be unwieldy as @chillu noted above. There is also a concern for reading times - not everyone can read that entire sentence in 10 seconds, and we have to think about accessibility concerns as well.

File: admin/javascript/SSNotice.js (hunk 3 above)
Comment (colymba) [inline admin/javascript/SSNotice.js]: This is just another way to create a P node with a class in jQuery. Since we are creating notices on the fly and each one is different, we can't really have them already in the template?

File: admin/javascript/LeftAndMain.js (hunk 2 above)
Comment (colymba) [inline admin/javascript/LeftAndMain.js]: Sorry, I'll put this back.

**Comment by colymba** [comment]
@chillu @stevie-mayhew How about leaving all notices and errors visible with an option/button to dismiss them? This will git people time to read them and the "jumping" will not be a problem since it would only happen when the notice is manually disabled.

My only concern about keeping "success notices" on until dismissed is that for example you get a "DataObject saved successfully" notice then you continue editing and the notice is still there even if there are now unsaved changes. We could here clean success notice when there are changes in the CMS form? This will bring back the ~30px jump, but wont be as much of an issue since the user will already have clicked or the user will already be editing. What do you think?

File: admin/javascript/SSNotice.js (hunk 3 above)
Comment (dhensby) [inline admin/javascript/SSNotice.js]: I understand this is creating a dom element through jQuery - but you're making some pretty fundamental decisions in the JS in regards to how they should be marked up.

Why should it always be a `<p>` and what if we want a different class (or extra ones)?

If you either have a JS template OR a hidden el which contains these items already, then you can pull them out a lot more nicely.

```js

$holderEl = $('.js-notice-holder');
$details = $holderEl.find('.js-error-details');

```

Then you're only relying on some abstract hooks to get the elements out and there is no mixing of concerns

Comment (colymba) [inline admin/javascript/SSNotice.js]: Fair enough. For classes, this is configurable through the `options.type`, similar thing could be done for the node type too.

Hidden el don't make sense to me since it would involve more DOM manipulation. I guess we could have a stripped down base notice in a JS template loaded via SS Requirements so dev should be able to override it if needed and it give us an easy place to edit the markup?

**Comment by stevie-mayhew** [comment]
I'd definitely prefer the option to have the user able to dismiss the notification. That way they are aware if they do not dismiss it and it remains around. I rarely dismiss anything though as usually a success means move to the next thing :wink:

File: admin/javascript/SSNotice.js (hunk 3 above)
Comment (dhensby) [inline admin/javascript/SSNotice.js]: Using `options.type` to add a set of custom classes is pretty bad; it seems like a way to shoe-horn style aspects into a element that is being created by JS.

I'm not sure that rendering and inserting a new node into the dom is less manipulation that just inserting text into an existing element; but it's beside the point.

We're working with an MVC framework and that should be core to our approach throughout our project. Here, we're clearly coupling our control with our view. Writing HTML in JS is no different to doing it in PHP and should be avoided as they are separate concerns.

**Comment by chillu** [comment]
> I think that an error or a notification should stay in place until the user dismisses it

Yeah, I don't think that's going to work with success notices - they don't *have* to be acknowledged. For warnings, it used to stick around if you hover over it, otherwise auto-remove after a few seconds. I think as a rule of thumb, few messages will be critical enough that we have to enforce user acknowledgement. From a high level, we should think: "How many clicks does a user need to achieve a certain outcome?".

>  We could here clean success notice when there are changes in the CMS form? This will bring back the ~30px jump, but wont be as much of an issue since the user will already have clicked or the user will already be editing.

I don't understand this, sorry. Could you rephrase please?

**Comment by stevie-mayhew** [comment]
@chillu I'm of the thought that a notice is there for a reason and you don't just automatically remove them. An analog is a post it note on your desk, you leave them there until you are done with them - you don't get 10  seconds to look at them or have to hover over them if you want them to stick around.

A user doesn't have to dismiss notifications, but they shouldn't be automatically dismissed either IMHO

File: admin/javascript/SSNotice.js (hunk 3 above)
Comment (colymba) [inline admin/javascript/SSNotice.js]: If we are inserting text into existing element we would have to duplicate those anyway to create new one etc...

Am all for moving the html to a template. Thought this could be done in a JS template a bit like https://github.com/silverstripe/silverstripe-framework/blob/3.2/javascript/UploadField_uploadtemplate.js so the template can easily be chosen when calling the JS function or overridden by dev if needed. How does that sound?

**Comment by colymba** [comment]
@chillu I meant that success notice could disappear when there are changes in the form. A bit like how the publish button changes colour when there are changes. Except here we remove a notice, so it doesn't still say "saved successfully" while there are new changes on the form... what do you think?

File: admin/javascript/SSNotice.js (hunk 3 above)
Comment (dhensby) [inline admin/javascript/SSNotice.js]: Sounds good :)

**Comment by clarkepaul** [comment]
Sorry I missed this thread. I'm not a fan of content jumping on the page because of the message, with the amount of messages some areas have they could get annoying as they stack up. 
For the notifications that don't require a response (like undo) I was envisioning we have something like a notification icon top right corner where all notifications move too after x time, but that this feed also acts like a recent timeline of activity, or if you missed the notice you have access to it.

**Comment by dhensby** [comment]
It's a real shame this hasn't got in after 2years.

@colymba are you still committed to help getting notification improvements in - I think it'd be of huge benefit.

However, it may now mean a lot of work as react is being brought in...

**Comment by colymba** [comment]
From what I'm reading now seems this is more going towards an overall redesign of the notification UI/UX.... so maybe this PR is outdated now...?

@dhensby it does seems this now means a lot of work, design and dev. and a lot changed for me, personally and professionally which takes up a lot of my time, so depending on when we want this in, I might not be able to do so much... it's tagged for 4.0.0... when would this be you think?

**Comment by dhensby** [comment]
4.0.0 stable likely to be some time in the next 6 months.

If it's not something you can commit to, that's fine and it can wait - there has been some work around notifications already (though nothing fantastic).

This may be something that @clarkepaul wants visual input on anyway and could be taken internally.

If you can't find the time, feel free to close this and I apologise that it's been open so long and never made it in - it's a loss to the project and it seems the blame has to lay with us :(

**Comment by clarkepaul** [comment]
As far as I'm aware I probably wont be doing any redesign of the notifications anytime soon as there is a huge list of things which seem to have more priority, also my ideas for what the notifications area could be haven't gone further than just ideas—just in case things get held back because of me.

**Comment by colymba** [comment]
@dhensby @clarkepaul since this is for 4.0.0 I should be able to work on something....

I'll close this PR since it is quite outdated now. I'll work on a new notifications redesign and tag you into it and we'll take it from there..... to be continued.

**Comment by dhensby** [comment]
Can't wait!
